### DISCUSSION TITLE
Speeding up "Sending Flush Table" with --less-locking?

### DISCUSSION THREAD

**Discussion by extreme-sysadminning** [mydumper/mydumper#1317] [Q&A]
Ahoi,

i have inherited a very busy database host with plenty of databases to backup. Think shared hosting scale, half a TB ram, 40 cores, hundreds of web projects, hundreds of databases. TBs of stuff.

While doing backups with mydumper (latest release, Servers are MariaDB 10.3, MySQL8 and MySQL5.7) we see two bottlenecks:

One is the single-threadedness of gzip, which is used to compress the resulting SQL-files. I think i will look into pigz to give it a bit more speed. Should be easy enough. :-)

The other is that Sending Flush Table takes a very long long time, 10, sometimes 15 minutes, in some cases:

```
2023-10-16 15:10:02 [INFO] - Acquiring DDL lock
2023-10-16 15:10:02 [INFO] - Acquiring Global lock
2023-10-16 15:10:02 [INFO] - Sending Flush Table
2023-10-16 15:20:02 [INFO] - Acquiring FTWRL
2023-10-16 15:20:03 [INFO] - Started dump at: 2023-10-16 15:20:03
```

Do you have any idea how to speed this up?
Would something like **--less-locking** help or will this work against our goal of consistent backups? I have looked at https://github.com/mydumper/mydumper/wiki/Locks, but i guess i better ask for your expertise.

I'm open for suggestions, even if it's a "live with it, it is what it is".

Regards,
Karsten

**Comment by viswaprsnl** [comment]
Using pigz should be a part of the flags @david-ducos-percona .gzip is becoming a big issue .

**Comment by Esysteme** [comment]
I am insterested by the answer also, because at end it's stay only big table and gzip slow down the backup. the idea when we allow 6 threads the goal will be to use free thread to increase compression with pigz

**Comment by davidducos** [comment]
Hi @Esysteme / @viswaprsnl / @extreme-sysadminning ,

Since v14, we have been working on removing using only gzip. We achieved it and implemented --exec-per-thread, now the compression is made by an external command, so you can use what ever you want to compress faster.
We keep the shortcut to GZIP and ZSTD with the -c option. Other shortcuts could be added in the future. 

About the locking mechanism that we use, first we need to understand what we are doing and what the FTWRL means for mydumper. 
We need to establish a point in time for all the threads, this means that ALL the threads HAVE TO executed the START TRANSACTION WITH CONSISTENT SNAPSHOT (STWCS) at the same point in time. In order to achieve it, the "only" option that we have is FTWRL as it stop all the writes. However, this is not 100% true, as if you enabled GTID and you use Percona Server, we can check an internal variable after executing the STWCS and check if all the threads are in sync. afaik, MySQL and MariaDB didn't implement yet. In order to use this functionality with Percona Server, you use --no-locks and check the log if it says that it is consistent or not.


Regards
